### DISCUSSION TITLE
Few questions about  adding support for the BSC network on a forked branch of Reth

### DISCUSSION THREAD

**Discussion by unclezoro** [paradigmxyz/reth#8162] [General]
hi, Reth team,I am a developer from the BNB Chain team. Our interest in the Reth project was piqued when we started working on the Akula project. The recent performance report and roadmap blog of Reth is quite impressive to us. As a result, we attempt to add support for the BSC network on a forked branch of Reth. We've encountered some high-level issues during our design process, and we'd like to consult with your team to ensure our implementation aligns with the best practices for modularity.

Here are the questions: 
1. At present, the 'stage sync' or the 'insert block' of blockchain trees are triggered by external programs such as the op-node or the consensus client. However, when we look at BSC or other EVM blockchains, the scenario is slightly different as they dont have these “sidecar” programs. Their block syncing is triggered by modules like the 'downloader' or the 'block fetcher' that sit atop the P2P network module. We could introduce a similar program that feeds the blockchain data, or we implement a different “downloader”. For the latter case, we didn't find any patterns to follow, nor did we know how to implement modular design in accordance with 'reth'，appreciate if any advice.
2. The implementation of Optimism currently does not support stage sync. Will there be future updates that enable stage sync for Optimism? When considering sync modes for other networks, which one we should refer to? Ethereum or Optimism?
3. The definition of consensus traits is rather simple. The “finalize” logic, which originally belonged to the consensus, has been shifted to the executor, as seen [here](https://github.com/ethereum/go-ethereum/blob/1a79f8fe58f9c42db1dcfd8599f386c8158069e8/consensus/beacon/consensus.go#L352). However, the “finalize” of BSC is quite heavy, check [here](https://github.com/bnb-chain/bsc/blob/43b2ffa63ba242437bcf19ee9bccf9595be0763a/consensus/parlia/parlia.go#L1144), we have to put it inside the “EVMProcessor”, make it quite fat. From the perspective of module cohesion, we hope that these logic can be incorporated into the consensus engine, but the consensus trait is so simple that we can hardly achieve this, any better idea?

**Comment by joshieDo** [comment]
re 2) As of yesterday it should be supported https://github.com/paradigmxyz/reth/pull/7552

**Reply by unclezoro** [reply]
Really good to know. Thanks @joshieDo

**Comment by mattsse** [comment]
hey @unclezoro , very excited about this and happy to help, feel free to ping me on tg @mattsse as well or a new channel.

for p2p we have this block import stub: https://github.com/paradigmxyz/reth/blob/d852f7f012c1811a3d1197fd9c5fc63be5ec8233/crates/net/network/src/manager.rs#L85-L86

which currently doesn't do anything in PoS, but this is how we envisioned block import for p2p would look like, happy to change things if that makes it more convenient for you.

bypassing the engine API would also be possible I assume, we made some progress on abstracting the launch process which might come in handy, but I'll probably need more context about your exact requirements, is there code I can look at?

**Reply by unclezoro** [reply]
thanks for the quick response. Good to know the `block_import ` stub, will check this.

Regarding the consensus engine, here is the POC code: 

`Finalize` in `EVMProcessor`:  https://github.com/bnb-chain/reth/blob/ed1475376fa138b315f86639df7135f0d59cad22/crates/revm/src/processor.rs#L682
`Parlia` consensus: https://github.com/bnb-chain/reth/tree/ed1475376fa138b315f86639df7135f0d59cad22/crates/consensus/parlia/src
BSC processor: https://github.com/bnb-chain/reth/blob/ed1475376fa138b315f86639df7135f0d59cad22/crates/revm/src/bsc/processor.rs

**Reply by unclezoro** [reply]
hi @mattsse  I can not ping you privately through TG because of your security setting I guess, this is my TG https://t.me/zorowhite, appreciate if you can add me as contact.
I think #2 and #3 is quite clear to us.  Regarding #1, the present auto seal consensus is limited to a solitary node. Would the team consider to extend it to a private network? Like a single auto seal miner is active, with other full nodes able to synchronize through p2p network. This could serve as a good example for us to broaden support to other networks like BSC, which does not have any consensus client or op-node.  

PS: [EIP225 Clique consensus protocol](https://eips.ethereum.org/EIPS/eip-225), if reth could implement such consensus, it will make things even simple.
